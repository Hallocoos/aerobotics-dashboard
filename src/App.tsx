import Header from "./navigation/Header";
import { Navigation } from "./navigation/Navigation";

export function App() {
	return (
		<div>
			<Header></Header>
			<Navigation></Navigation>
		</div>
	);
}