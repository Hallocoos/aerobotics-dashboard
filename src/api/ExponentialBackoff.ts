/* eslint-disable @typescript-eslint/no-explicit-any */
type RetryFunction = (...args: any[]) => Promise<any>;
type RetryOptions = {
	maxAttempts: number;
	baseDelayMs: number;
};

export async function fetchExponentialBackoff(fn: RetryFunction, options: RetryOptions = { maxAttempts: 5, baseDelayMs: 1000 }): Promise<any> {
	let attempt = 1;

	const execute = async (): Promise<any> => {
		try {
			return await fn();
		} catch (error) {
			if (attempt >= options.maxAttempts) {
				throw error;
			}

			const delayMs = options.baseDelayMs * Math.pow(2, attempt);
			console.debug(`Retry attempt ${attempt} after ${delayMs}ms`);
			await new Promise((resolve) => setTimeout(resolve, delayMs));

			attempt++;
			return execute();
		}
	};
	return execute();
}
