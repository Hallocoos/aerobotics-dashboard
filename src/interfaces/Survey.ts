import { Base } from "./Common";

export interface Surveys extends Base {
	results: Survey[];
}

export interface Survey {
	id: number;
	orchard_id: number;
	date: string;
	hectares: number;
	polygon: string;
}

export interface OrchardSurvey extends Base {
	results: TreeSurvey[];
}

export interface TreeSurvey {
	id: number;
	lat: number;
	lng: number;
	ndre: number;
	ndvi: number;
	volume: number;
	area: number;
	row_index: number;
	tree_index: number;
}