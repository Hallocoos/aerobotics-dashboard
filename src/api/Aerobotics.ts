import { Farms } from "../interfaces/Api";

const authToken = process.env.VITE_REACT_APP_AUTH_TOKEN;

const headers: HeadersInit = {
	"Accept": "application/json",
	"Authorization": authToken || ""
};

export const getFarms = async (): Promise<Farms> => {
	const result: Farms = await fetch(
		"/farming/farms",
		{
			method: "GET",
			headers: headers,
		})
		.then((response) => response.json());

	return result;
};