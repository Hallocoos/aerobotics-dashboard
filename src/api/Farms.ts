import { Bins, Farm, Farms, FruitSize } from "../interfaces/Farm";


const authToken = process.env.VITE_REACT_APP_AUTH_TOKEN;

const headers: HeadersInit = {
	"Accept": "application/json",
	"Authorization": authToken || ""
};

export const getAllFarms = async (): Promise<Farms> => {
	const result: Farms = await fetch(
		"/farming/farms",
		{
			method: "GET",
			headers: headers,
		})
		.then((response) => response.json());
	return result;
};

// TODO: Add params
export const getFarmById = async (farm_id: number): Promise<Farm> => {
	const result: Farm = await fetch(
		`/farming/farms/${farm_id}`,
		{
			method: "GET",
			headers: headers,
		})
		.then((response) => response.json());
	return result;
};

// TODO: Add params
export const getFarmBinsByFarmId = async (farm_id: number): Promise<Bins> => {
	const result: Bins = await fetch(
		`/farming/farms/${farm_id}/bins/`,
		{
			method: "GET",
			headers: headers,
		})
		.then((response) => response.json());
	return result;
};

// TODO: Add params
export const getFruitSizesByFarmId = async (farm_id: number): Promise<FruitSize> => {
	const result: FruitSize = await fetch(
		`/farming/farms/${farm_id}/fruit_sizes?`,
		{
			method: "GET",
			headers: headers,
		})
		.then((response) => response.json());
	return result;
};