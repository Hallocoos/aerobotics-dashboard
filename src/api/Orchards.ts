import { Orchard, Orchards } from "../interfaces/Orchard";

const authToken = process.env.VITE_REACT_APP_AUTH_TOKEN;

const headers: HeadersInit = {
	"Accept": "application/json",
	"Authorization": authToken || ""
};

// TODO: Add params
export const getAllOrchards = async (): Promise<Orchards> => {
	const result: Orchards = await fetch(
		"/farming/orchards/",
		{
			method: "GET",
			headers: headers,
		})
		.then((response) => response.json());
	return result;
};

// TODO: Add params
export const getOrchardById = async (id: number): Promise<Orchard> => {
	const result: Orchard = await fetch(
		`/farming/orchards/${id}/`,
		{
			method: "GET",
			headers: headers,
		})
		.then((response) => response.json());
	return result;
};

