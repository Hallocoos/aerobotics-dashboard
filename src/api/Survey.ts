import { OrchardSurvey, Surveys } from "../interfaces/Survey";

const authToken = process.env.VITE_REACT_APP_AUTH_TOKEN;

const headers: HeadersInit = {
	"Accept": "application/json",
	"Authorization": authToken || ""
};

// TODO: Add params
export const getAllSurveys = async (): Promise<Surveys> => {
	const result: Surveys = await fetch(
		"/farming/surveys/",
		{
			method: "GET",
			headers: headers,
		})
		.then((response) => response.json());
	return result;
};

// TODO: Add params
export const getSurveyById = async (survey_id: number): Promise<OrchardSurvey> => {
	const result: OrchardSurvey = await fetch(
		`/farming/surveys/${survey_id}/tree_surveys/`,
		{
			method: "GET",
			headers: headers,
		})
		.then((response) => response.json());
	return result;
};
