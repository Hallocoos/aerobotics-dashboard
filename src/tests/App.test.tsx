import { render, waitFor } from "@testing-library/react";
import { App } from "../App";
import { MemoryRouter } from "react-router-dom";
import * as farms from "../api/Farms";
import * as orchards from "../api/Orchards";
import * as survey from "../api/Survey";
import { allFarms } from "./mocks/Farms.mocks";
import { allOrchards } from "./mocks/Orchards.mocks";
import { allSurveys } from "./mocks/Surveys.mocks";

describe("App Tests", () => {
	it("Check that app renders correctly", async () => {
		jest.spyOn(farms, "getAllFarms").mockResolvedValue(allFarms);
		jest.spyOn(orchards, "getAllOrchards").mockResolvedValue(allOrchards);
		jest.spyOn(survey, "getAllSurveys").mockResolvedValue(allSurveys);

		const { findByTestId } = render(
			<MemoryRouter initialEntries={["/"]}>
				<App />
			</MemoryRouter>
		);
		const app = await waitFor(() => findByTestId(/app/i));

		expect(app).toMatchSnapshot();
	});
});