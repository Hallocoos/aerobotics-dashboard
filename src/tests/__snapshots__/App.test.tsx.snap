// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Tests Check that app renders correctly 1`] = `
<div
  class="app"
  data-testid="app"
>
  <div
    class="content"
  >
    <table
      class="table table-striped table-bordered table-hover"
    >
      <thead>
        <tr>
          <th>
            Farm Name
          </th>
          <th>
            Orchard name
          </th>
          <th>
            Total trees surveyed
          </th>
          <th>
            Latest survey date
          </th>
          <th>
            Details
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            farm 1
          </td>
          <td>
            wethwerthwerhwerh
          </td>
          <td>
            6
          </td>
          <td>
            Wed Jan 01 2020
          </td>
          <td>
            <button
              class="btn btn-primary"
              type="button"
            >
              <a
                href="/farm/1"
              >
                &gt; Details
              </a>
            </button>
          </td>
        </tr>
        <tr>
          <td>
            farm 1
          </td>
          <td>
            wherhwerwerhwerh
          </td>
          <td>
            6
          </td>
          <td>
            Wed Jan 01 2020
          </td>
          <td>
            <button
              class="btn btn-primary"
              type="button"
            >
              <a
                href="/farm/1"
              >
                &gt; Details
              </a>
            </button>
          </td>
        </tr>
        <tr>
          <td>
            farm 2
          </td>
          <td>
            xvbcxcvbxcvb
          </td>
          <td>
            3
          </td>
          <td>
            Sun Feb 02 2020
          </td>
          <td>
            <button
              class="btn btn-primary"
              type="button"
            >
              <a
                href="/farm/2"
              >
                &gt; Details
              </a>
            </button>
          </td>
        </tr>
        <tr>
          <td>
            farm 2
          </td>
          <td>
            xcvbxcv xcvb
          </td>
          <td>
            3
          </td>
          <td>
            Sun Feb 02 2020
          </td>
          <td>
            <button
              class="btn btn-primary"
              type="button"
            >
              <a
                href="/farm/2"
              >
                &gt; Details
              </a>
            </button>
          </td>
        </tr>
        <tr>
          <td>
            farm 3
          </td>
          <td>
            qazxxqazxqaz
          </td>
          <td>
            2
          </td>
          <td>
            Tue Mar 03 2020
          </td>
          <td>
            <button
              class="btn btn-primary"
              type="button"
            >
              <a
                href="/farm/3"
              >
                &gt; Details
              </a>
            </button>
          </td>
        </tr>
        <tr>
          <td>
            farm 3
          </td>
          <td>
            xqazzxqaqazx
          </td>
          <td>
            2
          </td>
          <td>
            Tue Mar 03 2020
          </td>
          <td>
            <button
              class="btn btn-primary"
              type="button"
            >
              <a
                href="/farm/3"
              >
                &gt; Details
              </a>
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
